src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -47,73 +82,145 @@ public SemanticTokensRangeHandler(IGlobalOptionService globalOptions)
+                // different compilation.
+                if (ChecksumIsUnchanged_NoLock(project, projectChecksum))
+                    return;
 
-                    return true;
+                if (!_projectIdToEventSource.TryGetValue(project.Id, out var eventSource))
+                {
+                    eventSource = new CompilationAvailableEventSource(_asyncListener);
+                    _projectIdToEventSource.Add(project.Id, eventSource);
                 }
-            }
 
-            var workspaceStatusService = project.Solution.Workspace.Services.GetRequiredService<IWorkspaceStatusService>();
-            var isFullyLoaded = await workspaceStatusService.IsFullyLoadedAsync(cancellationToken).ConfigureAwait(false);
+                eventSource.EnsureCompilationAvailability(project, () => OnCompilationAvailable(project, projectChecksum));
+            }
+        }
 
-            lock (_lock)
+        private void OnCompilationAvailable(Project project, Checksum projectChecksum)
+        {
+            lock (_gate)
             {
-                if (isFullyLoaded && !_projectIdToCompilation.ContainsKey(project.Id))
-                {
-                    // If the project's compilation isn't yet available, kick off a task in the background to
-                    // hopefully make it available faster since we'll need it later to compute tokens.
-                    var newCompilationTask = project.GetCompilationAsync(cancellationToken);
-                    _projectIdToCompilation.Add(project.Id, newCompilationTask);
-                }
+                // Paranoia: It's technically possible (though unlikely) for two calls to compute the compilation for
+                // the same project to come back and call into this.  This is because the
+                // CompilationAvailableEventSource uses cooperative cancellation to cancel the in-flight request before
+                // issuing the new one.  There is no requirement though that the inflight request actually stop (or run
+                // to completion) prior to the next request running and completing.  In practice this should not happen
+                // as cancellation is checked fairly regularly.  However, if it does, check and do not bother to issue a
+                // refresh in this case.
+                if (ChecksumIsUnchanged_NoLock(project, projectChecksum))
+                    return;
+
+                // keep track of this checksum.  That way we don't get into a loop where we send a refresh notification,
+                // then we get called back into, causing us to compute the compilation, causing us to send the refresh
+                // notification, etc. etc.
+                _projectIdToLastComputedChecksum[project.Id] = projectChecksum;
             }
 
-            return false;
+            // Enqueue an item to notify the client that they should do a refresh.
+            _semanticTokenRefreshQueue?.AddWork();
```

Comment:
nit: Should this perhaps be a contract check instead of `?` to ensure `_semanticTokenRefreshQueue` isn't null?

Reply:
Yup.  That's a good call.